### PR TITLE
Runtime configure shim

### PR DESCRIPTION
**Pull request by cpuguy83**
**- What I did**

Move all shim config into "runtime" type, and adds 2 new built-in runtimes to allow explicitly switching from runc+shimv1 and runc+shimv2

**- How I did it**

Add a `ShimConfig` type to `types.Runtime` which stores the shim config.
The daemon then uses that instead of having things spread between daemon and libcontainerd, some bool passing, etc.

This option is only exposed internally, i.e. it cannot be set from daemon.json.
This is due to issues being able to support unknown shims which requires knowing types to pass to those shims (e.g. does it need a runc v2 shim config, a runc v1 shim config, or some other entirely different shim). 

**- How to verify it**

Set `--default-runtime=com.docker.runtime.runc.v2`, run a container (and/or a plugin), check the process tree and see the container was created with `/usr/local/bin/containerd-shim-runc-v2`

**- Description for the changelog**

Add support new built-in runtime for using the containerd v2 shim by using the runtime `com.docker.runtime.runc.v2`

### CODE REVIEW

File: daemon/start_unix.go [hunk 1]
```diff
@@ -3,70 +3,20 @@
...
 )
 
-func (daemon *Daemon) getRuntimeScript(container *container.Container) (string, error) {
-	name := container.HostConfig.Runtime
-	rt := daemon.configStore.GetRuntime(name)
-	if rt == nil {
-		return "", errdefs.InvalidParameter(errors.Errorf("no such runtime '%s'", name))
-	}
-
-	if len(rt.Args) > 0 {
```
Comment (cpuguy83) [inline daemon/start_unix.go:24]: Note, I did not carry this over since I have no idea the purpose of it and it doesn't seem to be documented.

We can add it back, but I figured since it seems strange to begin with and probably no one knows the reason for it (if you do, let's document it better), just remove it. If it causes a regression we can fix it.

File: plugin/executor/containerd/containerd.go [hunk 2]
```diff
@@ -66,11 +67,8 @@ func deleteTaskAndContainer(ctx context.Context, cli libcontainerdtypes.Client,
 
 // Create creates a new container
 func (e *Executor) Create(id string, spec specs.Spec, stdout, stderr io.WriteCloser) error {
-	opts := runctypes.RuncOptions{
-		RuntimeRoot: filepath.Join(e.rootDir, "runtime-root"),
```
Comment (cpuguy83) [inline plugin/executor/containerd/containerd.go:70]: Note here I did not carry through with this custom runtime root path either since it doesn't really seem necessary.

File: daemon/runtime_unix.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	v2runcoptions "github.com/containerd/containerd/runtime/v2/runc/options"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/daemon/config"
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+)
+
+const (
+	defaultRuntimeName = "runc"
+	linuxV1Runtime     = "com.docker.runtime.runc.v1"
+	linuxV2Runtime     = "com.docker.runtime.runc.v2"
```
Comment (AkihiroSuda) [inline daemon/runtime_unix.go:19]: Why not just use containerd runtime name?

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: daemon/start_unix.go (hunk 1 above)
Comment (cpuguy83) [reply daemon/start_unix.go:24]: Yay I found the reason for this.

File: daemon/runtime_unix.go (hunk 3 above)
Comment (cpuguy83) [reply daemon/runtime_unix.go]: I did this because we provide our own configuration... but maybe it's not needed.

Comment (cpuguy83) [reply daemon/runtime_unix.go]: FYI this code moved in my last push, but is currently unchanged.

Comment (AkihiroSuda) [reply daemon/runtime_unix.go]: Let's just use containerd names to avoid confusion.

File: daemon/runtime_unix.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+		}
+	}()
+
+	for name, rt := range runtimes {
+		if len(rt.Args) == 0 {
+			continue
+		}
+
+		script := filepath.Join(tmpDir, name)
+		content := fmt.Sprintf("#!/bin/sh\n%s %s $@\n", rt.Path, strings.Join(rt.Args, " "))
```
Comment (AkihiroSuda) [inline daemon/runtime_unix.go:113]: ?

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: daemon/runtime_unix.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/daemon/config"
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/pkg/errors"
+)
+
+const (
+	defaultRuntimeName = "runc"
+
+	linuxShimV1 = "io.containerd.runtime.v1.linux"
```
Comment (AkihiroSuda) [inline daemon/runtime_unix.go:27]: error out if shim v1 is specified on cgroup2

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: daemon/runtime_unix.go (hunk 4 above)
Comment (thaJeztah) [reply daemon/runtime_unix.go:113]: You mean: why a script? I recall there was a discussion about that in https://github.com/moby/moby/pull/37978#issuecomment-436393528

File: daemon/config/config.go [hunk 6]
```diff
@@ -47,8 +44,22 @@ const (
...
 	// DefaultInitBinary is the name of the default init binary
 	DefaultInitBinary = "docker-init"
+
+	// StockRuntimeName is the reserved name/alias used to represent the
+	// OCI runtime being shipped with the docker daemon package.
+	StockRuntimeName = "runc"
+	// LinuxV1RuntimeName is the runtime used to specify runc+shimv1
+	LinuxV1RuntimeName = "com.docker.runtime.runc.v1"
+	// LinuxV2RuntimeName is the runtime used to specify runc+shimv2
+	LinuxV2RuntimeName = "com.docker.runtime.runc.v2"
```
Comment (thaJeztah) [inline daemon/config/config.go:54]: Should we create a `type` for this, so that they are more of an enum (and nicely grouped together in the Go Docs, ah-la https://godoc.org/github.com/docker/docker/api/types/mount#Consistency

File: daemon/config/config_common_unix.go [hunk 7]
```diff
@@ -29,6 +29,7 @@ type commonUnixBridgeConfig struct {
 func (conf *Config) GetRuntime(name string) *types.Runtime {
 	conf.Lock()
 	defer conf.Unlock()
+
```
Comment (thaJeztah) [inline daemon/config/config_common_unix.go:32]: unrelated change

File: libcontainerd/remote/client.go [hunk 8]
```diff
@@ -129,17 +127,13 @@ func (c *client) Restore(ctx context.Context, id string, attachStdio libcontaine
 	}, nil
 }
 
-func (c *client) Create(ctx context.Context, id string, ociSpec *specs.Spec, runtimeOptions interface{}, opts ...containerd.NewContainerOpts) error {
+func (c *client) Create(ctx context.Context, id string, ociSpec *specs.Spec, shim string, runtimeOptions interface{}, opts ...containerd.NewContainerOpts) error {
```
Comment (thaJeztah) [inline libcontainerd/remote/client.go:130]: Wondering if `shim` could be passed as one of the eixisting options (could it be in `opts` or in `runtimeOptions` 🤔

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/runtime_unix.go (hunk 4 above)
Comment (cpuguy83) [reply daemon/runtime_unix.go:113]: This is old code, I just moved it out of a huge file to group it with other runtime code.

It's here because containerd does not accept extra arguments passed to the oci runtime.

File: daemon/config/config.go [hunk 9]
```diff
@@ -47,8 +44,22 @@ const (
 	DisableNetworkBridge = "none"
 	// DefaultInitBinary is the name of the default init binary
 	DefaultInitBinary = "docker-init"
+
+	// StockRuntimeName is the reserved name/alias used to represent the
+	// OCI runtime being shipped with the docker daemon package.
+	StockRuntimeName = "runc"
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply daemon/config/config.go:50]: The variable name?

File: libcontainerd/remote/client.go (hunk 8 above)
Comment (cpuguy83) [reply libcontainerd/remote/client.go:130]: It would be nice, but the client seems to do some extra things with the runtime options (which apparently need to be passed on start in addition to create for the v2 runtime?), which need to be passed together.

File: daemon/runtime_unix.go (hunk 5 above)
Comment (cpuguy83) [reply daemon/runtime_unix.go:26]: Added.

File: daemon/config/config.go (hunk 6 above)
Comment (cpuguy83) [reply daemon/config/config.go:54]: Tried this, but it is painful to deal with. Opted to not do this.

File: daemon/runtime_unix.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/docker/pkg/ioutils"
+	"github.com/sirupsen/logrus"
+
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply daemon/runtime_unix.go]: Fixed.

File: daemon/config/config.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply daemon/config/config.go:50]: yes

**Comment by thaJeztah** [comment]
@AkihiroSuda @tonistiigi PTAL

**Comment by cpuguy83** [comment]
Updated this to add deprecation warnings for the v1 runtime.

Re: naming the runtimes

The containerd names are confusing. I'm not sure if it's good to use them or not. Just thinking about it.

**Comment by cpuguy83** [comment]
Tests are all green now, at least.

**Review by AkihiroSuda — APPROVED** [review]

**Review by fuweid — APPROVED** [review]
LGTM(nb)

**Comment by fuweid** [comment]
> The containerd names are confusing. I'm not sure if it's good to use them or not. Just thinking about it.

Legacy shim doesn't have version. Both runc.v1 and runc.v2 are shimv2 lol

**Comment by AkihiroSuda** [comment]
My concern is that runc.v1 with the com.docker. prefix is shim API v1...

**Comment by cpuguy83** [comment]
Yeah, perhaps I should just use the containerd shim names as the runtime name.
This is only included as a fallback for edge cases anyway, and we can document that.

**Comment by cpuguy83** [comment]
I've updated this to use the shim names as our runtime name.

**Review by AkihiroSuda — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

@tonistiigi PTAL

File: daemon/runtime_unix.go (hunk 4 above)
Comment (tonistiigi) [reply daemon/runtime_unix.go:112]: Unrelated to this PR but we should do a validation of the permissions of this path every time it is used. Otherwise, this looks quite scary.

**Review by tonistiigi — COMMENTED** [review]
SGTM

**Comment by thaJeztah** [comment]
Looks like this may have introduced a regression (panic when configuring custom runtime); https://github.com/docker/for-linux/issues/1169

File: daemon/daemon_unix.go [hunk 11]
```diff
@@ -729,55 +724,11 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
...
-
-func (daemon *Daemon) initRuntimes(runtimes map[string]types.Runtime) (err error) {
-	runtimeDir := filepath.Join(daemon.configStore.Root, "runtimes")
-	// Remove old temp directory if any
-	os.RemoveAll(runtimeDir + "-old")
-	tmpDir, err := ioutils.TempDir(daemon.configStore.Root, "gen-runtimes")
-	if err != nil {
-		return errors.Wrap(err, "failed to get temp dir to generate runtime scripts")
+	if hostConfig.Runtime == config.LinuxV1RuntimeName || (hostConfig.Runtime == "" && daemon.configStore.DefaultRuntime == config.LinuxV1RuntimeName) {
+		warnings = append(warnings, fmt.Sprintf("Configured runtime %q is deprecated and will be removed in the next release.", config.LinuxV1RuntimeName))
```
Comment (thaJeztah) [inline daemon/daemon_unix.go:728]: We need to add a line to the deprecation log; https://github.com/docker/cli/blob/master/docs/deprecated.md
